### PR TITLE
Add zip code field and slide in after card number > 5 digits

### PR DESCRIPTION
**Pull request by whenyournameiskyle**

### CODE REVIEW

**Comment by jxom** [comment]
Hey @klaymon, are you planning to re-open this? Seems like a useful feature! 😄

**Comment by whenyournameiskyle** [comment]
@jxom If you were interested! I accidentally opened this earlier 🙊

**Comment by jxom** [comment]
@klaymon Yeah, I'll check it out a bit later when I find some time! 
We don't really use ZIP codes for credit cards in Australia so I didn't really think about adding it in - but imagine it would be super helpful for countries/companies who need it! Thanks for raising this though. 🙏

File: src/index.js [hunk 1]
```diff
@@ -42,6 +45,7 @@ const InputWrapper = styled.label`
   margin-left: 0.5em;
   display: flex;
   align-items: center;
+  transition: transform 1s;
```
Comment (jxom) [inline src/index.js]: 👌

File: src/index.js [hunk 2]
```diff
@@ -100,7 +106,8 @@ type State = {
   cardImage: string,
   cardNumberLength: number,
   cardNumber: ?string,
-  errorText: ?string
+  errorText: ?string,
+  showZip: Boolean
```
Comment (jxom) [inline src/index.js]: The correct flow annotation for a boolean is `boolean`

File: src/index.js [hunk 3]
```diff
@@ -109,14 +116,17 @@ class CreditCardInput extends Component<Props, State> {
...
   static defaultProps = {
     cardCVCInputRenderer: inputRenderer,
     cardExpiryInputRenderer: inputRenderer,
     cardNumberInputRenderer: inputRenderer,
+    cardZipInputRenderer: inputRenderer,
     cardExpiryInputProps: {},
     cardNumberInputProps: {},
     cardCVCInputProps: {},
+    cardZipInputProps: {},
     cardImageClassName: '',
```
Comment (jxom) [inline src/index.js:134]: Can we perhaps have an extra prop called `enableZipInput`? 😄To cater for billing countries who don't require a zip code.

File: src/index.js [hunk 4]
```diff
@@ -187,6 +198,12 @@ class CreditCardInput extends Component<Props, State> {
       cardNumber
     });
 
+    if (cardNumberLength >= 6) {
+      this.setState({showZip: true})
```
Comment (jxom) [inline src/index.js]: Perhaps refactor to 

```javascript
this.setState({ showZip: cardNumberLength >= 6 })
```

?

File: src/index.js [hunk 5]
```diff
@@ -319,6 +336,49 @@ class CreditCardInput extends Component<Props, State> {
...
+  handleCardZipChange = (
+    { onChange }: { onChange?: ?Function } = { onChange: null }
+  ) => (e: SyntheticInputEvent<*>) => {
+    const zip = e.target.value;
+    const zipLength = zip.length;
+    const cardType = payment.fns.cardType(this.state.cardNumber);
+
+    this.setFieldValid();
+    if (zipLength >= 5) {
+      if (!isValidZip(zip, cardType)) {
```
Comment (jxom) [inline src/index.js]: Possibly refactor to:

```javascript
if (zipLength >= 5 && !isValidZip(zip, cardType)) {
  this.setFieldInvalid('Zip code is invalid');
}
```

Also, do we need `cardType` for `isValidZip`?

File: src/index.js [hunk 6]
```diff
@@ -366,6 +428,8 @@ class CreditCardInput extends Component<Props, State> {
       inputStyle,
       invalidStyle
     } = this.props;
+    const transformedStyle = showZip ? {transform: 'translateX(0)'} : {transform: 'translateX(4rem)'}
```
Comment (jxom) [inline src/index.js]: Might be better to move this into the `InputWrapper` styled component and pass `showZip` as a prop to it? I guess it reads better if the `transition` is next to it's target.

E.g.

```javascript
const InputWrapper = styled.label`
  ...
  transition: transform 1s;
  transform: translateX(${props => props.showZip ? '0' : '4rem' });
  ...
`;

<InputWrapper
  inputStyled={inputStyle}
  data-max="MM / YY 9"
  showZip={showZip}
>
```

**Review by jxom — CHANGES_REQUESTED** [review]
Looks great @klaymon! However, it seems that it doesn't auto-focus to the zip input field when I finish typing in the CVC (like it auto-focuses to expiry after card number, and auto-focuses to CVC after expiry). Would we be able to do this? 

Also, let me know if you don't have the time to make these changes. I can make these changes when I find the time! 🤓

File: src/index.js (hunk 3 above)
Comment (whenyournameiskyle) [reply src/index.js:134]: Would you prefer the default prop for `enableZipInput` to be false or true?

Comment (jxom) [reply src/index.js:134]: Hmm, lets leave it as false? ☺️

Comment (jxom) [reply src/index.js]: Hmm, lets leave it as false? ☺️

**Comment by jxom** [comment]
This is looking sweet, I’ll merge later today 👌

**Comment by jxom** [comment]
Published in `1.0.6`

tyvm

**Pull request merged**